File: backend/app/services/MovieRelationsService.py
from app.repositories.MovieRelationsRepository import MovieRelationsRepository
from app.services.database import db
from sqlalchemy.exc import SQLAlchemyError
from flask import current_app


class MovieRelationsService:
    def __init__(self):
        self.movie_relations_repository = MovieRelationsRepository(db.session)

    def add_actor_to_movie(self, movie_id, actor_id, role=None):
        try:
            return self.movie_relations_repository.add_actor_to_movie(
                movie_id, actor_id, role
            )
        except SQLAlchemyError as e:
            current_app.logger.error(f"Error adding actor to movie: {str(e)}")
            raise Exception("Failed to add actor to movie")

    def remove_actor_from_movie(self, movie_id, actor_id):
        try:
            return self.movie_relations_repository.remove_actor_from_movie(
                movie_id, actor_id
            )
        except SQLAlchemyError as e:
            current_app.logger.error(f"Error removing actor from movie: {str(e)}")
            raise Exception("Failed to remove actor from movie")

    def add_director_to_movie(self, movie_id, director_id):
        try:
            return self.movie_relations_repository.add_director_to_movie(
                movie_id, director_id
            )
        except SQLAlchemyError as e:
            current_app.logger.error(f"Error adding director to movie: {str(e)}")
            raise Exception("Failed to add director to movie")

    def remove_director_from_movie(self, movie_id, director_id):
        try:
            return self.movie_relations_repository.remove_director_from_movie(
                movie_id, director_id
            )
        except SQLAlchemyError as e:
            current_app.logger.error(f"Error removing director from movie: {str(e)}")
            raise Exception("Failed to remove director from movie")

    def add_genre_to_movie(self, movie_id, genre_id):
        try:
            return self.movie_relations_repository.add_genre_to_movie(
                movie_id, genre_id
            )
        except SQLAlchemyError as e:
            current_app.logger.error(f"Error adding genre to movie: {str(e)}")
            raise Exception("Failed to add genre to movie")

    def remove_genre_from_movie(self, movie_id, genre_id):
        try:
            return self.movie_relations_repository.remove_genre_from_movie(
                movie_id, genre_id
            )
        except SQLAlchemyError as e:
            current_app.logger.error(f"Error removing genre from movie: {str(e)}")
            raise Exception("Failed to remove genre from movie")

    def get_movie_actors(self, movie_id):
        try:
            movie_actors = self.movie_relations_repository.get_movie_actors(movie_id)
            return [
                {
                    "actor_id": ma.actor_id,
                    "actor_name": ma.actor.actor_name if hasattr(ma, "actor") else None,
                    "role": ma.movie_role,
                }
                for ma in movie_actors
            ]
        except SQLAlchemyError as e:
            current_app.logger.error(f"Error getting movie actors: {str(e)}")
            raise Exception("Failed to get movie actors")

    def get_movie_directors(self, movie_id):
        try:
            movie_directors = self.movie_relations_repository.get_movie_directors(
                movie_id
            )
            return [
                {
                    "director_id": md.director_id,
                    "director_name": (
                        md.director.director_name if hasattr(md, "director") else None
                    ),
                }
                for md in movie_directors
            ]
        except SQLAlchemyError as e:
            current_app.logger.error(f"Error getting movie directors: {str(e)}")
            raise Exception("Failed to get movie directors")

    def get_movie_genres(self, movie_id):
        try:
            movie_genres = self.movie_relations_repository.get_movie_genres(movie_id)
            return [
                {
                    "genre_id": mg.genre_id,
                    "genre_name": mg.genre.genre_name if hasattr(mg, "genre") else None,
                }
                for mg in movie_genres
            ]
        except SQLAlchemyError as e:
            current_app.logger.error(f"Error getting movie genres: {str(e)}")
            raise Exception("Failed to get movie genres")

File: filmhive-frontend/src/pages/Dashboard/services/movieRelationsService.ts
const getAuthToken = (): string | null => {
    return localStorage.getItem('accessToken');
};

export interface MovieActor {
    id: number;
    name: string;
    role?: string;
    photo_url?: string;
}

export interface MovieDirector {
    id: number;
    name: string;
    photo_url?: string;
}

export interface MovieGenre {
    genre_id: number;
    genre_name: string;
}

// Dodaj cache-busting headers
const getHeaders = (includeAuth: boolean = false) => {
    const headers: HeadersInit = {
        'Content-Type': 'application/json',
        'Cache-Control': 'no-cache, no-store, must-revalidate',
        'Pragma': 'no-cache',
        'Expires': '0'
    };

    if (includeAuth) {
        const token = getAuthToken();
        if (token) {
            headers['Authorization'] = `Bearer ${token}`;
        }
    }

    return headers;
};

// AKTORZY - używamy tego samego endpointu co w MovieDetails
export const addActorToMovie = async (movieId: number, actorId: number, role: string = ''): Promise<void> => {
    const token = getAuthToken();
    if (!token) {
        throw new Error('Brak tokenu autoryzacyjnego');
    }

    try {
        const response = await fetch(`http://localhost:5000/api/movie-relations/movies/${movieId}/actors`, {
            method: 'POST',
            headers: getHeaders(true),
            body: JSON.stringify({ actor_id: actorId, role })
        });

        if (!response.ok) {
            const errorData = await response.json();
            throw new Error(errorData.error || 'Nie udało się dodać aktora do filmu');
        }
    } catch (error) {
        console.error('Error adding actor to movie:', error);
        throw error;
    }
};

export const removeActorFromMovie = async (movieId: number, actorId: number): Promise<void> => {
    const token = getAuthToken();
    if (!token) {
        throw new Error('Brak tokenu autoryzacyjnego');
    }

    try {
        const response = await fetch(`http://localhost:5000/api/movie-relations/movies/${movieId}/actors/${actorId}`, {
            method: 'DELETE',
            headers: {
                'Authorization': `Bearer ${token}`,
                'Cache-Control': 'no-cache'
            }
        });

        if (!response.ok) {
            const errorData = await response.json();
            throw new Error(errorData.error || 'Nie udało się usunąć aktora z filmu');
        }
    } catch (error) {
        console.error('Error removing actor from movie:', error);
        throw error;
    }
};

// KLUCZOWA ZMIANA: używamy endpointu z MovieDetails który zwraca role
export const getMovieActors = async (movieId: number): Promise<MovieActor[]> => {
    try {
        // Używamy tego samego endpointu co w MovieDetails z include_roles=true
        const url = `http://localhost:5000/api/movies/${movieId}?include_roles=true&_t=${Date.now()}`;

        const response = await fetch(url, {
            headers: getHeaders()
        });

        if (!response.ok) {
            throw new Error('Nie udało się pobrać aktorów filmu');
        }

        const data = await response.json();
        // Zwracamy aktorów z rolami z głównego endpointu filmu
        return data.actors || [];
    } catch (error) {
        console.error('Error fetching movie actors:', error);
        throw error;
    }
};

// ALTERNATYWNIE: jeśli chcesz używać dedykowanego endpointu cast
export const getMovieActorsFromCast = async (movieId: number): Promise<MovieActor[]> => {
    try {
        const url = `http://localhost:5000/api/movies/${movieId}/cast?_t=${Date.now()}`;

        const response = await fetch(url, {
            headers: getHeaders()
        });

        if (!response.ok) {
            throw new Error('Nie udało się pobrać obsady filmu');
        }

        const data = await response.json();
        return data || [];
    } catch (error) {
        console.error('Error fetching movie cast:', error);
        throw error;
    }
};

// REŻYSERZY
export const addDirectorToMovie = async (movieId: number, directorId: number): Promise<void> => {
    const token = getAuthToken();
    if (!token) {
        throw new Error('Brak tokenu autoryzacyjnego');
    }

    try {
        const response = await fetch(`http://localhost:5000/api/movie-relations/movies/${movieId}/directors`, {
            method: 'POST',
            headers: getHeaders(true),
            body: JSON.stringify({ director_id: directorId })
        });

        if (!response.ok) {
            const errorData = await response.json();
            throw new Error(errorData.error || 'Nie udało się dodać reżysera do filmu');
        }
    } catch (error) {
        console.error('Error adding director to movie:', error);
        throw error;
    }
};

export const removeDirectorFromMovie = async (movieId: number, directorId: number): Promise<void> => {
    const token = getAuthToken();
    if (!token) {
        throw new Error('Brak tokenu autoryzacyjnego');
    }

    try {
        const response = await fetch(`http://localhost:5000/api/movie-relations/movies/${movieId}/directors/${directorId}`, {
            method: 'DELETE',
            headers: {
                'Authorization': `Bearer ${token}`,
                'Cache-Control': 'no-cache'
            }
        });

        if (!response.ok) {
            const errorData = await response.json();
            throw new Error(errorData.error || 'Nie udało się usunąć reżysera z filmu');
        }
    } catch (error) {
        console.error('Error removing director from movie:', error);
        throw error;
    }
};

export const getMovieDirectors = async (movieId: number): Promise<MovieDirector[]> => {
    try {
        const url = `http://localhost:5000/api/movies/${movieId}?_t=${Date.now()}`;

        const response = await fetch(url, {
            headers: getHeaders()
        });

        if (!response.ok) {
            throw new Error('Nie udało się pobrać reżyserów filmu');
        }

        const data = await response.json();
        // Zwracamy reżyserów z głównego endpointu filmu
        return data.directors || [];
    } catch (error) {
        console.error('Error fetching movie directors:', error);
        throw error;
    }
};

// GATUNKI
export const addGenreToMovie = async (movieId: number, genreId: number): Promise<void> => {
    const token = getAuthToken();
    if (!token) {
        throw new Error('Brak tokenu autoryzacyjnego');
    }

    try {
        const response = await fetch(`http://localhost:5000/api/movie-relations/movies/${movieId}/genres`, {
            method: 'POST',
            headers: getHeaders(true),
            body: JSON.stringify({ genre_id: genreId })
        });

        if (!response.ok) {
            const errorData = await response.json();
            throw new Error(errorData.error || 'Nie udało się dodać gatunku do filmu');
        }
    } catch (error) {
        console.error('Error adding genre to movie:', error);
        throw error;
    }
};

export const removeGenreFromMovie = async (movieId: number, genreId: number): Promise<void> => {
    const token = getAuthToken();
    if (!token) {
        throw new Error('Brak tokenu autoryzacyjnego');
    }

    try {
        const response = await fetch(`http://localhost:5000/api/movie-relations/movies/${movieId}/genres/${genreId}`, {
            method: 'DELETE',
            headers: {
                'Authorization': `Bearer ${token}`,
                'Cache-Control': 'no-cache'
            }
        });

        if (!response.ok) {
            const errorData = await response.json();
            throw new Error(errorData.error || 'Nie udało się usunąć gatunku z filmu');
        }
    } catch (error) {
        console.error('Error removing genre from movie:', error);
        throw error;
    }
};

export const getMovieGenres = async (movieId: number): Promise<MovieGenre[]> => {
    try {
        const url = `http://localhost:5000/api/movie-relations/movies/${movieId}/genres?_t=${Date.now()}`;

        const response = await fetch(url, {
            headers: getHeaders()
        });

        if (!response.ok) {
            throw new Error('Nie udało się pobrać gatunków filmu');
        }

        const data = await response.json();
        return data.genres || [];
    } catch (error) {
        console.error('Error fetching movie genres:', error);
        throw error;
    }
};

